src/core/geometry/qgsgeometryutils.cpp

```diff
@@ -1806,6 +1810,24 @@ void QgsGeometryUtils::weightedPointInTriangle( const double aX, const double aY
   pointY = rBy + rCy + aY;
 }
 
+bool QgsGeometryUtils::setMValueFromPoints( const QgsPointSequence &points, QgsPoint &point )
+{
+  bool rc = false;
+
+  for ( const QgsPoint &pt : points )
```

Comment:
Looking at the implementation, it doesn't actually do what I'd expect here. I'd have thought it would copy the m value of the coordinate from `points` with an m value which is **closest** to the original `point` x/y, not just the first one found. To me the "closest" behaviour would be more useful in a general case, as we could then potentially use it to restore m values for coordinates in geometries repaired using makevalid (and other geos based routines which strip m values). What do you think?

Reply:
Are you proposing to have this as an optional sub-routine is Fix Geometry for instance? By converting every vertex to points, and then joining the attributes back? I'm unsure how expensive this could be on big datasets.

Reply:
I'm not a heavy user of these functions but I agree with @nyalldawson, it looks strange to just pick the first point with a measure.

Reply:
> Looking at the implementation, it doesn't actually do what I'd expect here.

It works as defaultZValue works and used in many place where you have to add a Z value (mostly to add the first Z from a list of snapped points )

> I'd have thought it would copy the m value of the coordinate from `points` with an m value which is **closest** to the original `point` x/y, not just the first one found

+1 This is an idea I also have and want to push soon, but it's different from the need we have. 

> To me the "closest" behaviour would be more useful in a general case, as we could then potentially use it to restore m values for coordinates in geometries repaired using makevalid (and other geos based routines which strip m values). What do you think?

+1 too, I already had to write this kind of geos workaround, but never finalized and pushed into qgis. But it is also a different PR from this one

Reply:
Ok, that's fair enough. But in that case I'm not super-happy about the naming of this function. Something more explicit like "transferFirstMValueToPoint" would be better, just so there's absolutely no confusion about what this does. Not sure how to handle the existing z method though. Maybe just make the docs highlight the behaviour with a big "\warning ...." note.

Reply:
@nyalldawson Done. I've renamed (and depecated) setZValueFromPoints for transferFirstZValueToPoint